### PR TITLE
changes to custom model chunk upload api

### PR DESCRIPTION
**Pull request by rbhavna**
Signed-off-by: Bhavana Ramaram <rbhavna@amazon.com>

### Description
Created model meta API for uploading meta data. Made changes to existing chunk API so that user can upload chunks using modelID. Added limitations for chunk size to not be empty or exceed 10MB

Test coverage is low. Will add more tests later.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkAction.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+package org.opensearch.ml.common.transport.upload_chunk;
+
+
+import org.opensearch.action.ActionType;
+
+public class MLUploadModelChunkAction extends ActionType<MLUploadModelChunkResponse> {
+    public static MLUploadModelChunkAction INSTANCE = new MLUploadModelChunkAction();
+    public static final String NAME = "cluster:admin/opensearch/ml/upload_custom_model_chunk";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkAction.java:13]: We are going to support build-in models in future. How about changing to a more generic name `cluster:admin/opensearch/ml/upload_model_chunk`

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkInput.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        builder.startObject();
+        builder.field(MODEL_ID_FIELD, modelId);
+        builder.field(CHUNK_NUMBER_FIELD, chunkNumber);
+        builder.field(CONTENT_FIELD, content);
+        builder.endObject();
+        return builder;
+    }
+
+    public static MLUploadModelChunkInput parse(XContentParser parser, byte[] content) throws IOException {
+        Integer chunkNumber = -1;
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkInput.java:72]: Why set chunkNumber as "-1" here? Will you validate upload chunk input and throw exception ?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkRequest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             OutputStreamStreamOutput osso = new OutputStreamStreamOutput(baos)) {
+            actionRequest.writeTo(osso);
+            try (StreamInput input = new InputStreamStreamInput(new ByteArrayInputStream(baos.toByteArray()))) {
+                return new MLUploadModelChunkRequest(input);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to parse ActionRequest into MLTrainingTaskRequest", e);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkRequest.java:72]: Change `MLTrainingTaskRequest` to `MLUploadModelChunkRequest` in error message.

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import org.opensearch.ml.common.model.MLModelState;
+import org.opensearch.ml.common.model.TextEmbeddingModelConfig;
+import org.opensearch.ml.common.transport.upload.MLUploadInput;
+
+import java.io.IOException;
+
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+@Data
+public class MLUploadModelMetaInput implements ToXContentObject, Writeable{
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:28]: For model meta, seems the workd "upload" is not so proper as use don't upload any content, just create with some meta info. How about changing to `MLCreateModelMetaInput`

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    private String modelContentHash;
+    private MLModelConfig modelConfig;
+    private Integer totalChunks;
+
+    @Builder(toBuilder = true)
+    public MLUploadModelMetaInput(String name, FunctionName functionName, String version, String description, MLModelFormat modelFormat, MLModelState modelState, Long modelContentSizeInBytes, String modelContentHash, MLModelConfig modelConfig, Integer totalChunks) {
+        if (name == null) {
+            throw new IllegalArgumentException("model name is null");
+        }
+        if (functionName == null) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:60]: If functionName is not null, should set `this.functionName = functionName`

Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:60]: If functionName is not null, should set `this.functionName = functionName`

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            throw new IllegalArgumentException("model content hash value is null");
+        }
+        if (modelConfig == null) {
+            throw new IllegalArgumentException("model config is null");
+        }
+        if (totalChunks == null) {
+            throw new IllegalArgumentException("total chunks field is null");
+        }
+        this.name = name;
+        this.functionName = functionName;
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:79]: If `functionName` is null, this line will set function name as null.

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+import java.io.IOException;
+
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+@Data
+public class MLUploadModelMetaInput implements ToXContentObject, Writeable{
+
+    public static final String FUNCTION_NAME_FIELD = "function_name";
+    public static final String MODEL_NAME_FIELD = "name"; //mandatory
+    public static final String MODEL_VERSION_FIELD = "version"; //m
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:32]: Change `//m` to `//mandatory` ?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        } else {
+            out.writeBoolean(false);
+        }
+        if (modelState != null) {
+            out.writeBoolean(true);
+            out.writeEnum(modelState);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeLong(modelContentSizeInBytes);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:127]: use `writeOptionalLong`?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeInt(totalChunks);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (name != null) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:141]: Suggest remove the if null checking for mandatory fields as they never could be null. That means unit test can never cover the field ==  null case.

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaRequest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             OutputStreamStreamOutput osso = new OutputStreamStreamOutput(baos)) {
+            actionRequest.writeTo(osso);
+            try (StreamInput input = new InputStreamStreamInput(new ByteArrayInputStream(baos.toByteArray()))) {
+                return new MLUploadModelMetaRequest(input);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to parse ActionRequest into MLTrainingTaskRequest", e);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaRequest.java:72]: Change `MLTrainingTaskRequest` to `MLUploadModelMetaRequest` in error message.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    public void uploadModel(MLUploadModelChunkInput mlUploadInput, ActionListener<MLUploadModelChunkResponse> listener) {
+        final String modelId = mlUploadInput.getModelId();
+        // Check the size of the content not to exceed 10 mb
+        GetRequest getRequest = new GetRequest(ML_MODEL_INDEX).id(modelId);
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            client.get(getRequest, ActionListener.wrap(r -> {
+                log.info("Completed Get Model Request, id:{}", modelId);
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:52]: Change this log to debug level or remove it?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+            log.error(e.getMessage());
+            listener.onFailure(e);
+        }
+    }
+
+    public static String customModelId(String modelId, Integer chunkNumber) {
+        return modelId + "_" + chunkNumber;
+    }
+
+    private boolean isChunkExceeding10MB(final long length) {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:154]: We may change the size limit in future. How about changing to a more generic name like "validateChunkSize"

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        var isChunkExceedsSize = false;
+        double fileSizeInKB = length / 1024;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:156]: How about we keep the size limit consistent, use `ModelHelper.CHUNK_SIZE`?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaUploader.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Inject
+    public MLModelMetaUploader(MLIndicesHandler mlIndicesHandler, ThreadPool threadPool, Client client) {
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.threadPool = threadPool;
+        this.client = client;
+    }
+
+    public void uploadModelMeta(MLUploadModelMetaInput mlUploadModelMetaInput, ActionListener<String> listener) {
+        threadPool.executor(TASK_THREAD_POOL).execute(() -> {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaUploader.java:46]: Actualy model meta doc creation is light weight, we can run it directly, no need to put it into ML thread pool.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/TransportUploadModelChunkAction.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+@Log4j2
+public class TransportUploadModelChunkAction extends HandledTransportAction<ActionRequest, MLUploadModelChunkResponse> {
+    TransportService transportService;
+    ModelHelper modelHelper;
+    MLIndicesHandler mlIndicesHandler;
+    MLTaskManager mlTaskManager;
+    ClusterService clusterService;
+    ThreadPool threadPool;
+    Client client;
+    MLTaskDispatcher mlTaskDispatcher;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/TransportUploadModelChunkAction.java:38]: Remove unused fields like this one?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+                    listener.onFailure(e);
+                }
+            }));
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            listener.onFailure(e);
+        }
+    }
+
+    public static String customModelId(String modelId, Integer chunkNumber) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:150]: Can we rename the function name? Also add a comment why do we need this method.

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLUploadModelMetaAction.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+import org.opensearch.ml.common.transport.upload_chunk.MLUploadModelMetaRequest;
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.rest.action.RestToXContentListener;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+public class RestMLUploadModelMetaAction extends BaseRestHandler {
+    private static final String ML_UPLOAD_MODEL_META_ACTION = "ml_upload_model__meta_action";
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLUploadModelMetaAction.java:28]: `model__meta` -> is this intentional?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java (hunk 4 above)
Comment (rbhavna) [reply common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelMetaInput.java:28]: Sure. Will address it in next PR

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+                    try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, r.getSourceAsBytesRef())) {
+                        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                        // Use this model to update the chunk count
+                        MLModel existingModel = MLModel.parse(parser);
+                        existingModel.setModelId(r.getId());
+                        if (existingModel.getTotalChunks() <= mlUploadInput.getChunkNumber()) {
+                            throw new Exception("Chunk number exceeds total chunks");
+                        }
+                        byte[] bytes = mlUploadInput.getContent();
+                        if (bytes == null || bytes.length == 0 || isChunkExceeding10MB(bytes.length)) {
```
[the remark this answers is not in the document]
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:64]: Good point. Will address this in the next PR.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java (hunk 12 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:154]: Sure Yaliang. Makes more sense.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaUploader.java (hunk 14 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaUploader.java:46]: Sure. Will make the changes.

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLUploadModelMetaAction.java (hunk 17 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLUploadModelMetaAction.java:28]: No, it was by mistake. Will remove it in the next PR.

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkInput.java (hunk 2 above)
Comment (rbhavna) [reply common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkInput.java:72]: Its just a declaration actually. Will change it to null.

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLCreateModelMetaAction.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.common.transport.upload_chunk;
+
+import org.opensearch.action.ActionType;
+
+public class MLCreateModelMetaAction extends ActionType<MLCreateModelMetaResponse> {
+    public static MLCreateModelMetaAction INSTANCE = new MLCreateModelMetaAction();
+    public static final String NAME = "cluster:admin/opensearch/ml/upload_custom_model_meta";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLCreateModelMetaAction.java:12]: Change `upload_custom_model_meta`  as `create_model_meta`?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLCreateModelMetaRequest.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    public MLCreateModelMetaRequest(StreamInput in) throws IOException {
+        super(in);
+        this.mlCreateModelMetaInput = new MLCreateModelMetaInput(in);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException exception = null;
+        if (mlCreateModelMetaInput == null) {
+            exception = addValidationError("ML input can't be null", exception);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLCreateModelMetaRequest.java:48]: Change to "Model meta input can't be null"?

File: common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkAction.java (hunk 1 above)
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/upload_chunk/MLUploadModelChunkAction.java:13]: Update to `upload_model_chunk`?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+            }));
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            listener.onFailure(e);
+        }
+    }
+
+    public boolean validateChunkSize(final long length) {
+        var isChunkExceedsSize = false;
+        double fileSizeByLimit = length / ModelHelper.CHUNK_SIZE;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:153]: Who not just use `length > ModelHelper.CHUNK_SIZE`?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+            }, e -> {
+                if (e instanceof IndexNotFoundException) {
+                    listener.onFailure(new MLResourceNotFoundException("Failed to find model"));
+                } else {
+                    log.error("Failed to get ML model " + modelId, e);
+                    listener.onFailure(e);
+                }
+            }));
+        } catch (Exception e) {
+            log.error(e.getMessage());
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:146]: This way won't print exception stack trace, suggest change to `log.error("Fail to upload chunk for model " + modelId, e);`

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+    private final NamedXContentRegistry xContentRegistry;
+
+    @Inject
+    public MLModelChunkUploader(MLIndicesHandler mlIndicesHandler, Client client, final NamedXContentRegistry xContentRegistry) {
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    public void uploadModel(MLUploadModelChunkInput mlUploadInput, ActionListener<MLUploadModelChunkResponse> listener) {
```
[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:47]: Actually this is a good point. This long function will give you headache when you write tests. It's not easy to fully cover this function with many local variables and logic branches. At least the client.index(...) related part can be refactored into new functions.

[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:47]: Good point, if not easy or needs some time to rethink, I think we can refactor this when we enhance code coverage as this PR is already big

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLCreateModelMetaAction.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Creates a MLUploadModelMetaRequest from a RestRequest
+     *
+     * @param request RestRequest
+     * @return MLUploadModelMetaRequest
+     */
+    @VisibleForTesting
+    MLCreateModelMetaRequest getRequest(RestRequest request) throws IOException {
+        XContentParser parser = request.contentParser();
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLCreateModelMetaAction.java:59]: Have you checked the scenario when the API is not provided with a request body? I think we can add checking the request body like here in other APIs. https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/rest/RestMLProfileAction.java#L75. Also you can think about what can be processed without a RestRequest content.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:47]: Thanks @dhrubo-os  and @Zhangxunmt. I will look into it and try breaking this down into smaller functions

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLCreateModelMetaAction.java (hunk 24 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLCreateModelMetaAction.java:59]: Sure. Will add a condition to check the request body is not empty.

**Review by ylwu-amzn — DISMISSED** [review]

**Review by ylwu-amzn — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+                }
+            }));
+        } catch (Exception e) {
+            log.error("Fail to upload chunk for model " + modelId, e);
+            listener.onFailure(e);
+        }
+    }
+
+    public boolean validateChunkSize(final long length) {
+        var isChunkExceedsSize = false;
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:152]: why not `boolean`?

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaCreate.java [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+                    }, e -> {
+                        log.error("Failed to index model meta doc", e);
+                        listener.onFailure(e);
+                    }));
+                }, ex -> {
+                    log.error("Failed to init model index", ex);
+                    listener.onFailure(ex);
+                }));
+            } catch (Exception e) {
+                log.error("Failed to create model meta doc", e);
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelMetaCreate.java:81]: what does `Failed to init ` mean? Any difference from `Failed to create`? If not, Can we be consistent wth: `Failed to create ` which is easy to understand. Can please reflect it where it is applicable? IMO `Failed to init` or `Failed to index` is bit confusing in this context.

**Review by dhrubo-os — APPROVED** [review]
Please address 2 comments in the next PR. Thanks for these important changes.

File: plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java (hunk 25 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/upload_chunk/MLModelChunkUploader.java:152]: Thanks for pointing out. Anything should be fine here. But I am just used to declaring variables as var. Will declare appropriate type from next time.

**Pull request merged**